runtime/interpreter/variable.go

```diff
@@ -44,16 +65,59 @@ func (v *Variable) SetValue(interpreter *Interpreter, locationRange LocationRang
 
 var variableMemoryUsage = common.NewConstantMemoryUsage(common.MemoryKindVariable)
 
-func NewVariableWithValue(gauge common.MemoryGauge, value Value) *Variable {
+func NewVariableWithValue(gauge common.MemoryGauge, value Value) Variable {
 	common.UseMemory(gauge, variableMemoryUsage)
-	return &Variable{
+	return &SimpleVariable{
 		value: value,
 	}
 }
 
-func NewVariableWithGetter(gauge common.MemoryGauge, getter func() Value) *Variable {
+func NewVariableWithGetter(gauge common.MemoryGauge, getter func() Value) Variable {
 	common.UseMemory(gauge, variableMemoryUsage)
-	return &Variable{
+	return &SimpleVariable{
 		getter: getter,
 	}
 }
+
+type SelfVariable struct {
+	value   Value
+	selfRef ReferenceValue
+}
+
+var _ Variable = &SelfVariable{}
+
+func NewSelfVariableWithValue(interpreter *Interpreter, value Value) Variable {
+	common.UseMemory(interpreter, variableMemoryUsage)
+
+	semaType := interpreter.MustSemaTypeOfValue(value)
+
+	// Create an explicit reference to represent the implicit reference behavior of 'self' value.
+	// Authorization doesn't matter, we just need a reference to add to tracking.
+	selfRef := NewEphemeralReferenceValue(interpreter, UnauthorizedAccess, value, semaType, EmptyLocationRange)
+
+	return &SelfVariable{
+		value:   value,
+		selfRef: selfRef,
+	}
+}
```

Comment:
The actual fix is to introduce this special self variable, which internally hold an explicit reference to the actual value.
This reference is tracked and gets updated via usual reference tracking.
Then in the getter below, check for the reference validity before returning the value.